File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(begin(nums),end(nums));
        int n=nums.size();
        vector<vector<int>> res;
        for(int i=0;i<n-2;i++){
            // int x=nums[i];
            while(i>0 && i<n-2 && nums[i]==nums[i-1]) i++;
            if(i==n-2) break;
            int j=i+1,k=n-1;
            while(j<k){
                int sum=nums[i]+nums[j]+nums[k];
                if(!sum){
                    res.push_back({nums[i],nums[j++],nums[k--]});
                    while(j<k && nums[j]==nums[j-1]) j++;
                    while(j<k && nums[k]==nums[k+1]) k--;
                } 
                else if(sum>0) k--;
                else j++;
            }
        }

        return res;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums[:] = sorted(nums)
        n = len(nums)
        res = []
        for i in range(n): 
            while i>0 and i<n and nums[i]==nums[i-1]:
                i+=1
            j = i+1
            k = n-1 
            while j<k: 
                curr_sum = nums[i]+nums[j]+nums[k] 
                if curr_sum==0: 
                    res.append([nums[i],nums[j],nums[k]])
                    j+=1
                    k-=1
                    while j<k and nums[j]==nums[j-1]:
                        j+=1 
                    while j<k and nums[k]==nums[k+1]: 
                        k-=1
                elif curr_sum>0: 
                    k-=1 
                else: 
                    j+=1
        return res